### PR TITLE
update Windows build docs for Visual C++ 2015

### PR DESCRIPTION
**Pull request by zishiwu123**
Related to #1743 but for Visual C++ 2015 build tools, not Visual Studio IDE 2017-19. Enhancement to Windows build documentation.

**Before PR**
Windows build docs for Visual C++ 2015 previously instructed users who wanted to only download build tools to use [this link](https://visualstudio.microsoft.com/vs/older-downloads/#microsoft-build-tools-2015-update-3). However, downloading Microsoft Build Tools 2015 Update 3 option from that link is not enough. The installer wizard will fail and display and error message about missing dependencies without explicitly telling the user what was missing.

**After PR**
[Old link](https://visualstudio.microsoft.com/vs/older-downloads/#microsoft-build-tools-2015-update-3) has been replaced with a [new link](https://my.visualstudio.com/Downloads?q=visual%20studio%202015&wt.mc_id=o~msft~vscom~older-downloads) to download the Visual C++ Build Tools for Visual Studio 2015 with Update 3. The user only needs to login to their Microsoft account to access the downloads. The [Windows build tools documetation](https://github.com/DFHack/dfhack/blob/develop/docs/Compile.rst#microsoft-visual-studio-2015) has also been updated with instructions for users to download the build tools ISO file from new link because:
- The installer wizard inside the ISO file contains all the dependencies required by MSBuild v14.0, the version required by DFHack. Dependencies include:
  - Microsoft .NET Framework 4.6.1 Developer Pack
  - Microsoft Visual C++ 2015 Redistributable (x64) - 14.0.24210
  - Windows 10 Universal SDK - 10.0.10240
  - Windows 8.1 SDK
- The installer wizard is straightforward as the user only needs to stick to the default options and click next.
- Users who prefer getting the build tools as part of the Visual Studio IDE installations can still benefit from checking the dependencies listed in the build tools ISO file. It informs them what optional tools must be installed with the IDE to ensure that MSBuild v14.0 has all its dependencies met.

**Notes**
- I was unable to build DFHack for win64 until I changed the `build\win64\generate-MSVC-<options>.bat` files to use `v140` instead of `v140_xp` as [mentioned here](https://github.com/DFHack/dfhack/issues/1518#issuecomment-620938952). Then I was able to run `generate-MSVC-minimal.bat` and build DFHack.
- I was able to build DFHack for win32 using `v140_xp`.
- I found it easier to download just the build tools (v140) rather than download the Visual Studio IDE + build tools together. The build tools only was the only setup that worked for me, but [some people](https://github.com/DFHack/dfhack/issues/1518#issuecomment-622132767) have reported better luck with installing Visual Studio IDE + build tools together. Here are the 3 setups listed below:
  - Visual C++ 2015 build tools only
  - VS 2015 IDE + build tools (got ISO file for it from same link as the build tools only)
  - VS 2017 IDE + build tools (got from [here](https://my.visualstudio.com/Downloads?q=visual%20studio%202017&wt.mc_id=o~msft~vscom~older-downloads))
- The VS 2015 IDE installer also came in the form of an ISO file, but it was very slow to install and even slower to remove. It installed random things into my computer for mobile development and SQL Server, which I had to manually remove some folders because the installer wizard did not delete these folders even though I specified everything to be removed. It might be because I clicked the Windows 8.1 SDK option, which came with mobile app development tools. I saw that was a dependencies in the build tools only ISO file; hence why I clicked that option.
- The VS 2017 IDE installer downloaded MSBuild v15.0 which did not allow me to build because DFHack requires MSBuild v14.0. Here is the path to the MSBuild integrated with VS 2017 IDE: `C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\MSBuild\15.0\Bin`.

**Suggestions**:
- Update the `build\win64\generate-MSVC-<options>.bat` files to use `v140` instead of `v140_xp` as this error has already been reported in #1518. Not sure how many users of DFHack are running Windows XP as their main gaming platform. But I don't think there are enough that we need to have backwards compatibility with Windows XP at the cost of not being able to build for Windows 10. This opinion only applies to win64.
- [Some people](https://github.com/DFHack/dfhack/issues/1518#issuecomment-753334092) and are able to install newer versions of Visual Studio IDE (e.g. 2019) along with older build tools (MSBuild v14.0 which corresponds to the `v140` flag). But if someone does not already have VS2015 installed as [mentioned here](https://github.com/DFHack/dfhack/issues/1743#issuecomment-1068496970), I would advise them to install just the build tools like I did.

### CODE REVIEW

**Comment by ab9rf** [comment]
I'm going to reiterate that we should probably direct people to use VS 2019 or VS 2022, rather than VS 2017.

**Review by ab9rf — APPROVED** [review]
while this doesn't cover vs2019 or vs2022, it looks good for what it does cover

**Comment by zishiwu123** [comment]
I spent half my weekend trying to build DFHack on Windows. I'd motivated to improve the Windows build tools documentation so that developers using Windows to build DFHacks will not have to suffer so much.

File: docs/changelog.txt [hunk 1]
```diff
@@ -94,6 +94,9 @@ changelog.txt uses a syntax similar to RST, with a few special sequences:
 - ``argparse.processArgsGetopt()``: you can now have long form parameters that are not an alias for a short form parameter. For example, you can now have a parameter like ``--longparam`` without needing to have an equivalent one-letter ``-l`` param.
 - ``dwarfmode.enterSidebarMode()``: ``df.ui_sidebar_mode.DesignateMine`` is now a suported target sidebar mode
 
+## Documentation
```
Comment (myk002) [inline docs/changelog.txt:97]: Make sure you add this entry to the "Future" section above. Right now it's in the release notes for the last release (0.47.05-r4)

File: docs/Compile.rst [hunk 2]
```diff
@@ -485,11 +485,54 @@ other versions won't work against Dwarf Fortress due to ABI and STL incompatibil
 
 You can install Visual Studio 2015_ or 2017_ Community edition for free, which
 include all the features needed by DFHack. You can also download just the
-`build tools`_ if you aren't going to use Visual Studio to edit code.
+Visual C++ 2015 `build tools`_ if you aren't going to use Visual Studio to edit code.
+
+Option 1: Build Tools Only
+++++++++++++++++++++++++++
```
Comment (myk002) [inline docs/Compile.rst:491]: To fix the doc build warnings, try an underline style that hasn't been used in the file yet, like backticks (```````````)

**Comment by myk002** [comment]
Thank you for this PR! It does make things much clearer!

**Comment by ab9rf** [comment]
We can't use `v140_xp` instead of `v140` because that'll result in a binary that isn't ABI-compatible with Dwarf Fortress. DFHack has to be built with a compiler toolchain that is ABI-compatible with the one Toady uses, and since he has used the `v140_xp` toolchain for all recent builds, we have to as well.

You might be able to get DFHack to compile with `v140` but I'm not sanguine that it'll actually work when built that way.

**Comment by ab9rf** [comment]
An additional note: the `v141_xp` toolchain in VS2017 is reportedly ABI-compatible with the `v140_xp` toolchain of VS2015, and can probably be used to build DFHack successfully. This has not, however, been tested, to the best of my knowledge.

**Comment by zishiwu123** [comment]
> We can't use `v140` instead of `v140_xp` because that'll result in a binary that isn't ABI-compatible with Dwarf Fortress. DFHack has to be built with a compiler toolchain that is ABI-compatible with the one Toady uses to build Dwarf Fortress itself, and since he has used the `v140_xp` toolchain for all recent builds, we have to as well.
> 
> You might be able to get DFHack to compile with `v140` but I'm not sanguine that it'll actually work when built that way. `v140` and `v140_xp` have incompatible linkers.
> 
> Please confirm that the builds you created using this approach actually work with Dwarf Fortress.

Thank you for the explanation @ab9rf. How would you like me to test my build of DFHack with the `v140` toolchain? I am new to the game, a lurking fan who decided to finally try it out, so I am not familiar with all the features. So far I have created a new world and used the embark assistant plugin. I searched for 4x4 tiles with streams+ level rivers, dense level forest, and deep metals. The plugin found and highlighted matching areas.

Are there any other plugins or commands you would like me to test?

**Comment by ab9rf** [comment]
> Are there any other plugins or commands you would like me to test?

It's promising that you got that far without a crash, but I would like to see multiple fort-mode plugins tested, including those that create, modify, and destroy units and items. Also test as many plugins and scripts as possible which involve vmethod overrides, as those are highly likely to crash if there is an ABI incompatibility.

I have spent some time researching what the difference between `v140` and `v140_xp` and have not been able to find a clear statement as to what is different other than the linker. It's entirely possible that the compilers are the same and the only difference is in runtime linkage; if this is true, it should have little or no effect on DFHack itself because the parts of DFHack that have to be ABI compatible are only built to DLL and thus are not linked to a standalone executable.

Perhaps @quietust or one of our other developers knows more about the specifics of ABI compatibility of these different toolsets.

**Comment by zishiwu123** [comment]
The force push was because I accidentally typed the characters **`:** to the end of my commit message and wanted to remove those. I wasn't used to the key bindings on the default editor that comes with git for windows.
- fixed doc errors / moved changelog entry to future section`:

Question: for pull requests that have changes suggested by reviewers, do you prefer that: I place them in separate commits no matter how small the changes are? As opposed to running `git commit --amend` and then `git push --force` to overwrite a pushed commit?

I read the first option is safer than the second. Also I read from [this article](https://cloudfour.com/thinks/squashing-your-pull-requests/) that the commits in a PR can be squashed prior to merging.

**Comment by myk002** [comment]
I personally prefer squash and merge, keeping the related changes from one PR in one commit and setting the commit timestamp to the time of merge. This has the side benefit of not cluttering the main branch with merge commits.

If there are commits that I think should be kept separate, I use rebase and merge.

tl;dr we try to avoid a lot of merge commits in the main branch, but handle the commits in your PR however you like.

**Review by myk002 — APPROVED** [review]

**Comment by myk002** [comment]
@ab9rf should anything about v140 vs v140_xp be included in this PR, or is the text as written good for merge?

**Comment by zishiwu123** [comment]
@myk002 My intent in bringing up the `v140` vs `v140_xp` was not to add changes in the PR regarding this. I only wanted to bring this to your attention again since it was mentioned in #1518.

I think this PR should only focus on updating the Windows build documentation, specifically on how to install build tools. I think other matters such `v140` vs `v140_xp` should be opened in a separate issue so we don't clutter the scope of this PR.

**Comment by myk002** [comment]
Looking through the issues list, this PR is also relevant to #1480

**Comment by ab9rf** [comment]
> @ab9rf should anything about v140 vs v140_xp be included in this PR, or is the text as written good for merge?

We need better information on the difference between `v140` and `v140_xp` before we make a recommendation in this area. Until we have that information, I don't want to put anything into the documentation on this issue.

**Comment by myk002** [comment]
> We need better information on the difference between `v140` and `v140_xp` before we make a recommendation in this area. Until we have that information, I don't want to put anything into the documentation on this issue.

Ack. I'll merge this PR, then, and we can update the recommendation on the build chain once you're research is complete.

**Comment by ab9rf** [comment]
I've completed a build using both `v140_xp` and `v140`. While there are differences between the resulting DLLs, as far as I can tell they are immaterial. Some data areas are shuffled, but all the same data is present, and the code generation appears identical. As DFHack (and probably also Dwarf Fortress itself) doesn't run reliably on 64-bit XP even when built with `v140_xp`, due to a bug in the MSFT-provided libraries used with that build chain that MSFT has never fixed, there should be no harm in using `v140` instead of `v140_xp` as the build toolchain, at least on 64-bit platforms. However, we should continue to do release builds using `v140_xp`, especially for the 32-bit version, which still works fine on XP.

**Comment by myk002** [comment]
@ab9rf could you add that into the build docs? Maybe also a comment would be useful in https://github.com/DFHack/dfhack/blob/develop/build/win64/generate-MSVC-all.bat

**Comment by ab9rf** [comment]
> @ab9rf could you add that into the build docs? Maybe also a comment would be useful in https://github.com/DFHack/dfhack/blob/develop/build/win64/generate-MSVC-all.bat

#2034
